src/uct/cuda/cuda_copy/cuda_copy_md.c

```
@@ -336,15 +330,119 @@ static void uct_cuda_copy_sync_memops(uct_cuda_copy_md_t *md,
                                   (CUdeviceptr)address));
 }
 
+static ucs_status_t uct_cuda_copy_push_ctx(CUdevice orig_device,
+                                           CUdevice device,
+                                           ucs_log_level_t log_level)
+{
+    ucs_status_t status;
+    CUcontext primary_ctx;
+
+    if (orig_device == device) {
+        return UCS_OK;
+    }
+
+    status = UCT_CUDADRV_FUNC(cuDevicePrimaryCtxRetain(&primary_ctx, device),
```

I think it is better to check the primary device context using `cuDevicePrimaryCtxGetState`. The main use case that should be covered when user allocates memory on the device `I`, then switch to device `J`, and we should allocate staging buffer on device `I`. Thus, if `cuDevicePrimaryCtxGetState` returns inactive state for the given device, then we should report an error instead of retaining and activating primary context on the device that it has not been done yet.

---

added function, and log

---

i assume this code will be modified to use common func for context retainig, so no need to review it now, right?

---

removed the two function above and using `uct_cuda_primary_ctx_retain()`